File: leetcode/28.find-the-index-of-the-first-occur-in-a-string/cpp/strstr.cpp
#include <string>
#include <vector>

class Solution
{
public:
    int strStr(std::string haystack, std::string needle)
    {
        if (needle.empty())
            return 0;

        auto prefixLengths = getNextArr(needle);

        int j = 0;
        for (int i = 0; i < haystack.size(); ++i)
        {
            while (j > 0 && haystack[i] != needle[j])
                j = prefixLengths[j - 1];

            if (haystack[i] == needle[j])
                ++j;

            if (j == needle.size())
                return i - needle.size() + 1;
        }

        return -1;
    }

private:
    std::vector<int> getNextArr(std::string s)
    {
        std::vector<int> prefixLengths(s.size());
        prefixLengths[0] = 0;

        int j = 0;
        for (int i = 1; i < prefixLengths.size(); ++i)
        {
            while (j > 0 && s[i] != s[j])
                j = prefixLengths[j - 1];

            if (s[i] == s[j])
                ++j;

            prefixLengths[i] = j;
        }

        return prefixLengths;
    }
};

File: leetcode/28.find-the-index-of-the-first-occur-in-a-string/ts/strstr.ts
const strStr = (haystack: string, needle: string): number => {
    const next = getNextArr(needle)

    let j = 0
    for (let i = 0; i < haystack.length; i++) {
        while (j > 0 && haystack[i] !== needle[j]) j = next[j - 1]

        if (haystack[i] === needle[j]) j++

        if (j === needle.length) return i - needle.length + 1
    }

    return -1
}

const getNextArr = (s: string): number[] => {
    const next: number[] = Array.from({ length: s.length })
    next[0] = 0

    let j = 0
    for (let i = 1; i < s.length; i++) {
        while (j > 0 && s[i] !== s[j]) j = next[j - 1]

        if (s[i] === s[j]) j++

        next[i] = j
    }

    return next
}

export default strStr
